Pandaproxy support in operator

dimitriscruz opened the pull request:
## Cover letter

Adding support for configuring pandaproxy (and client) through the Redpanda operator. This initial support follows a similar pattern as in Kafka and Admin API - up to 1 internal listener, up to 1 external listener, and use of nodeports. The external connectivity can only be enabled if it is in the Kafka API and the subdomains need to be the same. These conditions, among others, are caught by the webhook.

Example of generated`redpanda.yaml` with external connectivity and use of subdomain:

```
redpanda:
   ...
    kafka_api:
        - address: 0.0.0.0
          port: 9092
          name: kafka
        - address: 0.0.0.0
          port: 9093
          name: kafka-external
    advertised_kafka_api:
        - address: cluster-proxy-0.cluster-proxy.default.svc.cluster.local.
          port: 9092
          name: kafka
        - address: 0.bla.com
          port: 32423
          name: kafka-external
```

```
pandaproxy:
    pandaproxy_api:
        - address: 0.0.0.0
          port: 8082
          name: proxy
        - address: 0.0.0.0
          port: 8083
          name: proxy-external
    advertised_pandaproxy_api:
        - address: cluster-proxy-0.cluster-proxy.default.svc.cluster.local.
          port: 8082
          name: proxy
        - address: 0.bla.com
          port: 30546
          name: proxy-external
pandaproxy_client:
    brokers:
        - address: cluster-proxy-0.cluster-proxy.default.svc.cluster.local.
          port: 9092
```

## Release notes

Cluster type in Kubernetes now supports pandaproxy. The user can enable the pandaproxy by adding an entry in the Cluster CR. Example:

```
  configuration:
   ...
    pandaproxyApi:
    - port: 8082
    - external:
        enabled: true
        subdomain: "test.subdomain.com"
    kafkaApi:
    - port: 9092
    - external:
        enabled: true
        subdomain: "test.subdomain.com"
```

senior7515 commented:
🔥  amazing.

alenkacz commented on src/go/k8s/apis/redpanda/v1alpha1/cluster_types.go:
```diff
@@ -138,9 +138,10 @@ type ClusterStatus struct {
+	Internal           []string `json:"internal,omitempty"`
+	External           []string `json:"external,omitempty"`
+	ExternalAdmin      []string `json:"externalAdmin,omitempty"`
+	ExternalPandaproxy []string `json:"externalPandaproxy,omitempty"`
```
should we call it just ExternalProxy?

alenkacz replied:
I guess if we go that way, it should be in all places. I just find it very long and confusing that it's one word :D

dimitriscruz replied:
It's "pandaproxy" in the docs and (cpp) source code, so I thought let's follow that..

alenkacz commented on src/go/k8s/apis/redpanda/v1alpha1/cluster_types.go:
```diff
@@ -166,10 +167,12 @@ type ClusterList struct {
+	KafkaAPI         []KafkaAPI       `json:"kafkaApi,omitempty"`
+	AdminAPI         []AdminAPI       `json:"adminApi,omitempty"`
+	PandaproxyAPI    []PandaproxyAPI  `json:"pandaproxyApi,omitempty"`
+	PandaproxyClient PandaproxyClient `json:"pandaproxy_client,omitempty"`
```
json: pandaproxyClient

dimitriscruz replied:
I removed the client field

alenkacz commented on src/go/k8s/apis/redpanda/v1alpha1/cluster_types.go:
```diff
@@ -258,6 +276,18 @@ type AdminAPITLS struct {
 }
 
+// PandaproxyAPITLS TLS settings for Pandaproxy
+type PandaproxyAPITLS struct {
```
could we share the two structs here?

dimitriscruz replied:
I removed anything TLS from this PR (types, webhook).

alenkacz commented on src/go/k8s/pkg/resources/resource.go:
```diff
@@ -34,6 +34,10 @@ const (
 	// ExternalListenerName is name of external listener
 	ExternalListenerName = "kafka-external"
+	// PandaproxyPortInternalName is the name of the pandaproxy internal port
+	PandaproxyPortInternalName = "proxy"
```
nice 👏

alenkacz commented on src/go/k8s/controllers/redpanda/cluster_controller.go:
```diff
@@ -65,6 +65,7 @@ type ClusterReconciler struct {
 //
 // For more details, check Reconcile and its Result here:
 // - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.7.0/pkg/reconcile
+// nolint:funlen // todo break down
```
:D

alenkacz commented on src/go/k8s/apis/redpanda/v1alpha1/cluster_webhook.go:
```diff
@@ -203,6 +207,69 @@ func (r *Cluster) validateKafkaListeners() field.ErrorList {
+		}
+		if proxyExternal != nil {
+			allErrs = append(allErrs,
+				field.Invalid(field.NewPath("spec").Child("configuration").Child("pandaproxyApi"),
```
I think the field should also include index: `.Child("pandaproxyApi").Index(i)`

dimitriscruz replied:
Added

alenkacz commented on src/go/k8s/apis/redpanda/v1alpha1/cluster_webhook.go:
```diff
@@ -203,6 +207,69 @@ func (r *Cluster) validateKafkaListeners() field.ErrorList {
+		proxyExternal = &r.Spec.Configuration.PandaproxyAPI[i]
+		if proxyExternal.Port != 0 {
+			allErrs = append(allErrs,
+				field.Invalid(field.NewPath("spec").Child("configuration").Child("pandaproxyApi"),
```
Index also here

dimitriscruz replied:
Added

alenkacz commented on src/go/k8s/apis/redpanda/v1alpha1/cluster_webhook.go:
```diff
@@ -203,6 +207,69 @@ func (r *Cluster) validateKafkaListeners() field.ErrorList {
+			allErrs = append(allErrs,
+				field.Invalid(field.NewPath("spec").Child("configuration").Child("pandaproxyApi"),
+					r.Spec.Configuration.PandaproxyAPI,
+					"cannot have a pandaproxy external listener without a kafka external listener"))
```
hmm... how come? why is this not possible?

alenkacz commented on src/go/k8s/apis/redpanda/v1alpha1/cluster_webhook.go:
```diff
@@ -203,6 +207,69 @@ func (r *Cluster) validateKafkaListeners() field.ErrorList {
+					r.Spec.Configuration.PandaproxyAPI,
+					"cannot have a pandaproxy external listener without a kafka external listener"))
+		}
+		if kafkaExternal == nil && proxyExternal.External.Subdomain != "" {
```
is this correct? So there's no kafka external connectivity but the error message is saying that the subdomains don't match

dimitriscruz replied:
The proxy specifies a subdomain while Kafka is empty, so the subdomains don't match.

I modified the text so it's more direct.

alenkacz commented on src/go/k8s/apis/redpanda/v1alpha1/cluster_webhook.go:
```diff
@@ -203,6 +207,69 @@ func (r *Cluster) validateKafkaListeners() field.ErrorList {
+	}
+
+	// TLS for pandaproxy is not supported yet
+	for i, p := range r.Spec.Configuration.PandaproxyAPI {
```
why don't we just drop it from the CRD then and add it when it can be configured?

dimitriscruz replied:
Yeah, dropped.

alenkacz commented on src/go/k8s/apis/redpanda/v1alpha1/cluster_webhook.go:
```diff
@@ -203,6 +207,69 @@ func (r *Cluster) validateKafkaListeners() field.ErrorList {
+		}
+	}
+
+	if !((len(r.Spec.Configuration.PandaproxyAPI) == 2 && proxyExternal != nil) || (proxyExternal == nil && len(r.Spec.Configuration.PandaproxyAPI) <= 1)) {
```
uff it's hard to parse these conditions :D not your fault, just noting it here if you can come up with easier way to put this in. Maybe split it into two named functions? 🤔

dimitriscruz replied:
Agree, I'll leave this for a separate PR for all listeners.

alenkacz commented on src/go/k8s/apis/redpanda/v1alpha1/cluster_webhook.go:
```diff
@@ -203,6 +207,69 @@ func (r *Cluster) validateKafkaListeners() field.ErrorList {
+		allErrs = append(allErrs,
+			field.Invalid(field.NewPath("spec").Child("configuration").Child("pandaproxyApi"),
+				r.Spec.Configuration.PandaproxyAPI,
+				"up to one internal listener and no external listener, or one external and one internal listener for pandaproxy is allowed"))
```
is internal listener "up to one" or exactly one? Because I think in some places in code you expect that internal listener exists (e.g. when computing external port - you expect that internal port exists as well and external is internal+1)

dimitriscruz replied:
"up to one". The callers of `PandaproxyAPIInternal()` check for nil.

These are the valid cases for pandaproxy ports:
```
(internal, external)
0,0
1,0
1,1
```

alenkacz commented on src/go/k8s/apis/redpanda/v1alpha1/cluster_webhook.go:
```diff
@@ -300,25 +381,31 @@ func (r *Cluster) ValidateDelete() error {
 	return nil
 }
 
+// nolint:funlen,gocyclo // this function needs rewriting once stabilized
```
oh gosh, this is the worst part of the codebase right now I think, I am refactoring it once this merges :D It was such a pain to handle even in my previous PRs

dimitriscruz replied:
Agree.

alenkacz commented on src/go/k8s/apis/redpanda/v1alpha1/cluster_webhook_test.go:
```diff
@@ -279,6 +279,38 @@ func TestValidateUpdate_NoError(t *testing.T) {
 		assert.Error(t, err)
 	})
+
+	t.Run("proxy subdomain must be the same as kafka subdomain", func(t *testing.T) {
```
I think you need to add the tests to both Create and Update.

dimitriscruz replied:
Thanks, added.

alenkacz commented on src/go/k8s/cmd/configurator/main.go:
```diff
@@ -40,20 +40,24 @@ const (
+	configDestination         string
+	nodeName                  string
+	subdomain                 string
+	externalConnectivity      bool
```
this is not for this PR but we should change this, there's no global external connectivity anymore so the name of this at least internally should change

dimitriscruz replied:
Agree. Also "listener" across the codebase could become "kafkaAPIListerner" like other APIs.

alenkacz commented on src/go/k8s/cmd/configurator/main.go:
```diff
@@ -40,20 +40,24 @@ const (
+	redpandaRPCPort           int
+	hostPort                  int
+	proxyExternalConnectivity bool
+	proxyHostPort             int
```
if we pull the nodeport service from here, is it also problematic as pulling the CRD? (just asking since you said you tried it :))

dimitriscruz replied:
It works fine for core resources and CRDs. We need a couple of additions in the cluster role the operator creates.

I have a working branch with this approach if we ever need to go that route.

alenkacz commented on src/go/k8s/cmd/configurator/main.go:
```diff
@@ -174,28 +218,65 @@ func registerAdvertisedKafkaAPI(
+	}
+
+	// Pandaproxy uses the Kafka API subdomain.
+	if len(c.subdomain) > 0 {
```
what if pandaproxy does not specify subdomain but kafka does? that's also not allowed? and is it the case for admin api as well?

alenkacz commented on src/go/k8s/apis/redpanda/v1alpha1/cluster_types.go:
```diff
@@ -166,10 +167,12 @@ type ClusterList struct {
+	KafkaAPI         []KafkaAPI       `json:"kafkaApi,omitempty"`
+	AdminAPI         []AdminAPI       `json:"adminApi,omitempty"`
+	PandaproxyAPI    []PandaproxyAPI  `json:"pandaproxyApi,omitempty"`
+	PandaproxyClient PandaproxyClient `json:"pandaproxy_client,omitempty"`
```
can we dropped this since it's not used and add it when supported?

alenkacz commented on src/go/k8s/pkg/resources/statefulset.go:
```diff
@@ -480,6 +480,21 @@ func overprovisioned(developerMode bool, limits corev1.ResourceList) []string {
+	listener := r.pandaCluster.PandaproxyAPIExternal()
+	if listener != nil {
+		envs = append(envs, corev1.EnvVar{
+			Name:  "PROXY_EXTERNAL_CONNECTIVITY",
```
do we need this? We already add PandaproxyPortExternalName listener to the configmap so can't we use that to figure out that there's external connectivity instead of introducing another env var?

dimitriscruz replied:
No, I was on the fence on this one because we already follow this pattern for the Kafka listener.

I removed the bool and infer it instead.

alenkacz reviewed (changes requested):
Huh, pretty big effort, nice work 👏 

Looks overall good, I have just couple of comments around validation rules && dropping unsupported fields since that's pretty much dead code (I don't see any advantage of adding it now vs adding it later - adding fields is not breaking change)

alenkacz reviewed (changes requested):
Thanks for all the additions. If not having internal is valid case, I don't think this would work 

https://github.com/vectorizedio/redpanda/pull/1307/files#diff-208e6df471016ebc997b72c36741150b86b9023d2d8d698bbcb3a8e56f0f84d2R113

That's the last standing issue from my side then 👍

alenkacz commented:
we agreed with @dimitriscruz that we'll make internal listener mandatory for now to solve that so when that commit lands, feel free to approve and merge 👍

dimitriscruz commented:
> we agreed with @dimitriscruz that we'll make internal listener mandatory for now to solve that so when that commit lands, feel free to approve and merge 👍

Thanks, Alena! I added the check in the webhook to prevent CRs with 1 external and 0 internal pandaproxy listeners. Also added this scenario in the webhook tests.

alenkacz reviewed (approved):
Looks great, thanks
